compiler/mir/mirpasses.nim

```diff
@@ -277,8 +313,174 @@ proc preventRvo(tree: MirTree, changes: var Changeset) =
+
+    result = true # until proven otherwise
+    for i in 1..min(an.len, bn.len):
+      let a {.cursor.} = tree[an[^i]]
+      let b {.cursor.} = tree[bn[^i]]
+
+      template check(cond: bool) =
+        if cond:
+          # bail out once the paths diverge
+          result = false
+          break
+
+      check(a.kind != b.kind)
+
+      case a.kind
+      of SymbolLike:
+        check(a.sym.id != b.sym.id)
+      of mnkTemp:
+        check(a.temp != b.temp)
+      of mnkPathNamed, mnkPathVariant:
+        check(a.field.id != b.field.id)
+      of mnkPathPos:
+        check(a.position != b.position)
+      of mnkPathArray:
+        check(sameIndex(tree[an[^i] - 3], tree[bn[^i] - 3]) == no)
+      else:
+        unreachable()
+
+  # we follow op-params in the analysis, meaning that we can ignore
+  # ``mnkRegion``s here
+  for i in search(tree, {mnkCall}):
+    # regions are allowed to not use an arg-block, and we don't want/need to
+    # analyse those that don't
+    if not isArgBlock(tree, i - 1):
+      continue
+
+    var iter = initArgIter(tree, i)
+    while iter.hasNext:
+      let arg = rawNext(iter, tree)
+      case tree[arg].kind
+      of mnkArg:
+        # check if a value that is potentially the same as the analysed
+        # one is:
+        # - potentially mutated while evaluating following arguments
+        # - passed to a parameter of the analysed call that allows for
+        #   mutations
+        # If either is the case, we need to introduce a shallow copy and use
+        # that as the argument
+        let val = skipTag(tree, unaryOperand(tree, Operation arg))
+
+        # 1. an r-value is unique, meaning that we know that a temporary is
+        #    not needed
+        # 2. aliasing is preferred over stack overflows
+        # XXX: ^^ it shouldn't be. A static error would be better...
+        if isRvalue(tree, getRoot(tree, val).OpValue) or
+           getSize(config, tree[val].typ) >= 1024:
+          continue
+
+        var needsTemp = false
+        block:
+          # first, check whether the argument value is passed to an earlier
+          # ``var`` parameter of the procedure (e.g., given the call
+          # ``f(a, b, c, d)`` and analysing 'c', check 'a' and 'b'. 'd' is
+          # analysed as part of the ``potentialMutations`` loop below)
+          var iter2 = iter
+          while iter2.hasNext:
+            let arg2 = rawNext(iter2, tree)
+            if tree[arg2].kind == mnkName and tree[arg2 - 1].kind == mnkTag:
+              if maybeSameMutableLocation(tree, val, OpValue(arg2 - 2)):
+                needsTemp = true
+                break
+
+        if not needsTemp:
+          # look for potential mutations of the argument that happen after
+          # the argument is bound
+          for mut in potentialMutations(tree, arg + 1, i - 2):
+            if maybeSameMutableLocation(tree, val, mut):
+              needsTemp = true
+              break
```

Comment:
We could move both loops into a block and break out of that common block. I think the current way is quite readable too though.
```suggestion
        block checkIfArgNeedsTemp:
          # first, check whether the argument value is passed to an earlier
          # ``var`` parameter of the procedure (e.g., given the call
          # ``f(a, b, c, d)`` and analysing 'c', check 'a' and 'b'. 'd' is
          # analysed as part of the ``potentialMutations`` loop below)
          var iter2 = iter
          while iter2.hasNext:
            let arg2 = rawNext(iter2, tree)
            if tree[arg2].kind == mnkName and tree[arg2 - 1].kind == mnkTag:
              if maybeSameMutableLocation(tree, val, OpValue(arg2 - 2)):
                needsTemp = true
                break checkIfArgNeedsTemp

          # look for potential mutations of the argument that happen after
          # the argument is bound
          for mut in potentialMutations(tree, arg + 1, i - 2):
            if maybeSameMutableLocation(tree, val, mut):
              needsTemp = true
              break checkIfArgNeedsTemp
```

Reply:
Hm, yeah, that's probably cleaner. With the separate `if` statement, I think my intention was to make it clear that `iter2` is only meant to be used with the first loop, but the code is short enough for this to be visible without the separate scope.